### ISSUE TITLE
Can accumulate be called after combine for reductions?

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#16459]
I'm looking at bug in arkouda's mink/maxk that was made more likely with https://github.com/chapel-lang/chapel/pull/16401, but I can repro with 1.22 as well.

The maxk code uses a heap to keep track of the max k values while accumulating, and then as an optimization when combining it un-heapifys by sorting the buffer and merges sorted buffers. However, I'm seeing some additional calls to accumulate after that sorting, and the data structure is no longer suitable for the heaped accumulate so it's getting wrong answers. I wasn't sure if accumulate after combine was legal or not so wanted to ask here.

Here's a reproducer (just took the technote reduction and added halts if accumulate  is called after combine)

```chpl
class PlusReduceOp: ReduceScanOp {
  type eltType;
  var value: eltType;
  var startedCombine = false;

  proc identity {
    return 0: eltType;
  }

  proc accumulate(elm) {
    if startedCombine then halt("accum after combine");
    value = value + elm;
  }

  proc accumulateOntoState(ref state, elm) {
    if startedCombine then halt("accum after combine");
    state = state + elm;
  }

  proc initialAccumulate(outerVar) {
    value = value + outerVar: eltType;
  }

  proc combine(other: borrowed PlusReduceOp) {
    startedCombine = true;
    value = value + other.value;
  }

  proc generate() return value;

  proc clone() return new unmanaged PlusReduceOp(eltType=eltType);
}

var A = [1000, 200, 30, 4];
var sum: int;
forall elm in A with (PlusReduceOp reduce sum) {
  sum += elm;  // equivalently:  sum reduce= elm;
}
writeln(sum);
```

**Comment by vasslitvinov** [comment]
Turns out sometimes we have to `accumulate` after `combine`. Consider this standalone iterator used with a `+` reduction:
```chpl
iter simple(param tag==standalone) {
  coforall 1..1 do
    yield 2;

  yield 3;
}

var sum = + reduce simple();
```

The top-level AS (accumulation state) needs to `combine` the results coming from the child tasks, if any. In this case, it is a single task producing `2`.

Since the parallel iterator executes a top-level `yield` after completing the child task, the top-level AS needs to `accumulate` the corresponding value, in this case `3`. This happens after a `combine`.

In the arkouda case I suspect that the parallel iterators do not have top-level yields. So there is a possibility to eliminate the top-level accumulate-after-combine, similar to https://github.com/chapel-lang/chapel/issues/16366 .

It is implicit in the previous paragraph that the current implementation creates a top-level shadow variable "just in case there is a top-level yield". It is initialized to identity and them accumulate-d into the top-level accumulation state. When there are no top-level yields, this is a no-op and could be avoided.

**Comment by bradcray** [comment]
Given Vass's diagnosis of the Arkouda case, I wonder whether the min-max heap that Arkouda is using in those ops could simply identity accumulates and drop them on the ground; and to assert if it gets a non-identity accumulate after it's already switched to "combine" mode.

**Comment by vasslitvinov** [comment]
> Could the top-level yield(s?) accumulate into a new, fresh state and then have that state combine into the top-level one?

Potentially, yes. Part of our push in #9879 is to make the generated code more efficient, avoiding unnecessary copying. We can come back to this when implementing #9879.

**Comment by bradcray** [comment]
> Could the top-level yield(s?) accumulate into a new, fresh state and then have that state combine into the top-level one?

I was thinking we could potentially combine this effort with #16366 in that the additional reduction state for yields would only be created, used, and combined in the event that there are top-level yields (which I think is a rare case in general, particularly for our most important / high-performance iterators).

**Comment by ronawho** [comment]
> Given Vass's diagnosis of the Arkouda case, I wonder whether the min-max heap that Arkouda is using in those ops could simply identity accumulates and drop them on the ground; and to assert if it gets a non-identity accumulate after it's already switched to "combine" mode.

Yeah, this is kind of what I'm going for. I was going to make it work even in this case, but not worry about efficiency.

**Comment by vasslitvinov** [comment]
> I was going to make it work even in this case, but not worry about efficiency.

My two cents about this direction is that it would present a hidden performance trap that the user would need to figure out.

**Comment by ronawho** [comment]
>> I was going to make it work even in this case, but not worry about efficiency.

> My two cents about this direction is that it would present a hidden performance trap that the user would need to figure out.

Ok, we ended up converting an assert.
